actors/builtin/miner/vesting_state.go

```diff
@@ -0,0 +1,125 @@
+package miner
+
+import (
+	"sort"
+
+	"github.com/filecoin-project/specs-actors/actors/abi"
+	"github.com/filecoin-project/specs-actors/actors/abi/big"
+)
+
+// VestingFunds represents the vesting table state for the miner.
+// It is a slice of (VestingEpoch, VestingAmount).
+// The slice will always be sorted by the VestingEpoch.
+type VestingFunds struct {
+	Funds []VestingFund
+}
+
+func (v *VestingFunds) unlockVestedFunds(currEpoch abi.ChainEpoch) abi.TokenAmount {
+	amountUnlocked := abi.NewTokenAmount(0)
+
+	lastIndexToRemove := -1
+	for i, vf := range v.Funds {
+		if vf.Epoch >= currEpoch {
+			break
+		}
+
+		amountUnlocked = big.Add(amountUnlocked, vf.Amount)
+		lastIndexToRemove = i
+	}
+
+	// remove all entries upto and including lastIndexToRemove
+	if lastIndexToRemove != -1 {
+		v.Funds = v.Funds[lastIndexToRemove+1:]
+	}
+
+	return amountUnlocked
+}
+
+func (v *VestingFunds) addLockedFunds(currEpoch abi.ChainEpoch, vestingSum abi.TokenAmount,
+	provingPeriodStart abi.ChainEpoch, spec *VestSpec) {
+	// maps the epochs in VestingFunds to their indices in the slice
+	epochToIndex := make(map[abi.ChainEpoch]int, len(v.Funds))
+	for i, vf := range v.Funds {
+		epochToIndex[vf.Epoch] = i
+	}
+
+	// Quantization is aligned with when regular cron will be invoked, in the last epoch of deadlines.
+	vestBegin := currEpoch + spec.InitialDelay // Nothing unlocks here, this is just the start of the clock.
+	vestPeriod := big.NewInt(int64(spec.VestPeriod))
+	vestedSoFar := big.Zero()
+	for e := vestBegin + spec.StepDuration; vestedSoFar.LessThan(vestingSum); e += spec.StepDuration {
+		vestEpoch := quantizeUp(e, spec.Quantization, provingPeriodStart)
+		elapsed := vestEpoch - vestBegin
+
+		targetVest := big.Zero() //nolint:ineffassign
+		if elapsed < spec.VestPeriod {
+			// Linear vesting, PARAM_FINISH
+			targetVest = big.Div(big.Mul(vestingSum, big.NewInt(int64(elapsed))), vestPeriod)
+		} else {
+			targetVest = vestingSum
+		}
+
+		vestThisTime := big.Sub(targetVest, vestedSoFar)
+		vestedSoFar = targetVest
+
+		// epoch already exists. Load existing entry
+		// and update amount.
+		if index, ok := epochToIndex[vestEpoch]; ok {
+			currentAmt := v.Funds[index].Amount
+			v.Funds[index].Amount = big.Add(currentAmt, vestThisTime)
+		} else {
+			// append a new entry -> slice will be sorted by epoch later.
+			entry := VestingFund{Epoch: vestEpoch, Amount: vestThisTime}
+			v.Funds = append(v.Funds, entry)
+			epochToIndex[vestEpoch] = len(v.Funds) - 1
+		}
+	}
+
+	// sort slice by epoch
+	sort.Slice(v.Funds, func(first, second int) bool {
+		return v.Funds[first].Epoch < v.Funds[second].Epoch
+	})
+}
+
+func (v *VestingFunds) unlockUnvestedFunds(currEpoch abi.ChainEpoch, target abi.TokenAmount) abi.TokenAmount {
+	amountUnlocked := abi.NewTokenAmount(0)
+	lastIndexToRemove := -1
+
+	for i, vf := range v.Funds {
+		if amountUnlocked.GreaterThanEqual(target) {
+			break
+		}
+
+		if vf.Epoch >= currEpoch {
```

Comment:
This is still an issue. We can't just skip the funds that _should_ have already been unlocked and drop them on the floor.

We either need to:

1. Admit that this case should be impossible, and return an error if it does happen for some reason.
2. Even better, just handle it and unlock the "should have already been unlocked" funds.

Reply:
Good catch! This wasn't an issue in the original, which left the leading AMT keys in place. Could we please have a test for this explicitly, and then option 2.

Reply:
On the options:
1. It seems quite possible; 2 out of 3 call paths to here have not processed vesting immediately prior
2. This will require passing out the unlocked amount and correctly using it to update pledge with the power actor

So I suggest:
3. Replicate the prior behaviour of leaving the should-have-vested entries in the table.

Reply:
This is done using option 3 and a test has been added for this.